execution_engine/src/storage/trie/merkle_proof.rs

```diff
@@ -0,0 +1,323 @@
+                let (indexed_pointers_with_hole, rem): (Vec<(u8, Pointer)>, &[u8]) =
+                    FromBytes::from_bytes(rem)?;
+                Ok((
+                    TrieMerkleProofStep::Node {
+                        hole_index,
+                        indexed_pointers_with_hole,
+                    },
+                    rem,
+                ))
+            }
+            TRIE_MERKLE_PROOF_STEP_EXTENSION_ID => {
+                let (affix, rem): (Vec<u8>, &[u8]) = FromBytes::from_bytes(rem)?;
+                Ok((TrieMerkleProofStep::Extension { affix }, rem))
+            }
+            _ => Err(bytesrepr::Error::Formatting),
+        }
+    }
+}
+
+/// A proof that a node with a specified `key` and `value` is present in the Merkle trie.
+/// Given a state hash `x`, one can validate a proof `p` by checking `x == p.compute_state_hash()`.
+#[derive(Debug, PartialEq, Eq)]
+pub struct TrieMerkleProof<K, V> {
+    key: K,
+    value: V,
+    proof_steps: VecDeque<TrieMerkleProofStep>,
+}
+
+impl<K, V> TrieMerkleProof<K, V> {
+    /// Constructor for [`TrieMerkleProof`]
+    pub fn new(key: K, value: V, proof_steps: VecDeque<TrieMerkleProofStep>) -> Self {
+        TrieMerkleProof {
+            key,
+            value,
+            proof_steps,
+        }
+    }
+
+    /// Getter for the value in [`TrieMerkleProof`]
+    pub fn value(&self) -> &V {
+        &self.value
+    }
+}
+
+impl<K, V> TrieMerkleProof<K, V>
+where
+    K: ToBytes + Copy,
+    V: ToBytes + Copy,
+{
+    /// Recomputes a state root hash from a [`TrieMerkleProof`].
+    /// This is done in the following steps:
+    ///
+    /// 1. Using [`TrieMerkleProof::key`] and [`TrieMerkleProof::value`], construct a
+    /// [`Trie::Leaf`] and compute a hash for that leaf.
+    ///
+    /// 2. We then iterate over [`TrieMerkleProof::proof_steps`] left to right, using the hash from
+    /// the previous step combined with the next step to compute a new hash.
+    ///
+    /// 3. When there are no more steps, we return the final hash we have computed.
+    ///
+    /// The steps in this function reflect `operations::rehash`.
+    pub fn compute_state_hash(&self) -> Result<Blake2bHash, Error> {
+        let mut hash = {
+            let leaf_bytes = Trie::leaf(self.key, self.value).to_bytes()?;
+            Blake2bHash::new(&leaf_bytes)
+        };
+
+        for (proof_step_index, proof_step) in self.proof_steps.iter().enumerate() {
+            let pointer = if proof_step_index == 0 {
+                Pointer::LeafPointer(hash)
+            } else {
+                Pointer::NodePointer(hash)
+            };
+            let proof_step_bytes = match proof_step {
+                TrieMerkleProofStep::Node {
+                    hole_index,
+                    indexed_pointers_with_hole,
+                } => {
+                    let hole_index = *hole_index;
+                    if hole_index as usize > RADIX {
```

Comment:
But this will be always `true` as long as `hole_index` is defined as `u8`? And it's a bit weird that `proof_step_index` is also `u8` in the error enum. I don't have full context on this change but from what I understand previously we used `usize` for indices, but now u8 - wouldn't it possibly cause more unintended side effects?

Reply:
You are right.  Thing we should change this to an assert, just in case `RADIX` changes?

Reply:
I will add that changing indices to `u8` is necessary because we now need to be serializing them for client consumption.

Reply:
FIXED